Everything (not meant to got in, just for testing, will split into more PRs)

Filoppi opened the pull request:
-New audio mixer, capable of playing samples at any speed or sample rate, also has a lower latency overall.
-Audio now slows down when falling behind the target speed by "Audio Emu Speed Tolerance"
-New audio stretcher implementation, rarely ever crackles and is way more responsive than before.
-New WASAPI implementation, fixed all the crashes, error checks/leaks and some compatibility problems
-New mouse capture setting for when playing in a Window
-Fixes for game window accepting inputs after it lost focus
-Improved mouse axis accuracy, you can now play GC FPS games with a mouse easily
-Loads of new input macro functions (some to play FPS games with a mouse)
-Restored custom input modifiers (they were stuck at 0.5 before)
-Added settings on inputs to still pass them when we have no focus (e.g. battery level or extensions)
-Fixed FPS counter and game window speed breaking on pause/unpause
-Audio pane UI fixes
-Misc fixes

iwubcode commented on Source/Core/InputCommon/ControlReference/ControlReference.cpp:
```diff
@@ -4,19 +4,59 @@
+  HasFocus = 0x01,
+  // Full focus means the cursor is captured/locked inside the game window.
+  // Ignored if mouse capturing is off. Implies HasFocus
+  HasFullFocus = 0x02,
```
Just going off the comment, maybe better as `RestrictMouseToWindow` ?

Filoppi replied:
It's not exactly that, when you don't have FullFocus other inputs can be ignored if you want, not only the mouse.
But yes, mainly, it's for the mouse. I know, it's an ambiguous name for an abstract concept, but I think the name explains it fine.

Filoppi commented:
I would be greateful if a few people could test the new audio stretching, it's 98% done.
Both on weak and powerful PCs and on Android.
Do not test GC games as I think I broke something about the GC sample rate, still testing that, Super Mario Sunshine works fine though. Wii games work.
I need to find the best default value for a setting, which is the new "Emulation Speed Tolerance" in the Audio Panel.
It is how much time the audio needs to fall behind the target emulation speed before falling back to the ACTUAL emulation speed itself.
By target emulation speed I mean, the emulation speed the emulation tries to reach, whether 100%, 10% or 200%...
By actual emulation speed I mean the speed at which the emulation is actually going, which is calculated by the audio mixer, counted from DMA samples pushes as had the best balance between number of pushes per second and accuracy.

A value of about 10ms will immediately make the audio playback speed fallback to the actual emulation speed if there is a smaller stutter,
which might be undesirable if you don't have sound stretching on, especially because it happens when you alt tab and stuff...
But given that I have fixed the loss in quality from keeping the stretcher off, you might keep it on at all times now.
Values too high will cause crackling in the first ms of a stutter, but it should never be as bad as before, for 2 reasons:
When speed keeps going down, audio is slowed down dramatically so new sounds are always played, instead of padded silence,
while is some other occasions, the last samples you had will back playwards, filling up the time with samples that sound good.
If you are debugging and you don't want to be bothered with the audio being linked to the actual elapsed time, you can just disable it by sliding the slider left.

Another thing to test is Config::MAIN_AUDIO_MIXER_LATENCY, which is set to 40ms, targeting half of it, so 20ms.
The same latency is used by stretching or non stretching.
40ms seems to be fine and I had no problems with it, but other machines might. It's customizable from the ini.

Filoppi commented:
**WASAPI changes:**
Fix: WASAPI using wrong latency (device min latency + user selected latency, instead of the max of the 2)
Feature: WASAPI now works with 5.1/DPLII. It would be very easy to add support for any new audio formats now.
Feature: Added support to any sample rate to WASAPI (user selected, some are restricted but can be easily allowed).
Fix: WASAPI not initializing correctly in multiple cases, mostly related to latency.
Fix: Some WASAPI COM objects not being released.
Fix: WASAPI improved errors handling and log printing.
Fix: WASAPI crashing Dolphin if we changed the properties of the used device while a game was running, it now keeps trying to restarts WASAPI until Windows has finished reconfiguring it.
Fix: WASAPI stopping the game thread for the duration of the backend latency when the emulation was paused or stopped.
Fix: WASAPI not being fully thread safe. Crashing when pausing/unpausing, or stopping.
Fix: WASAPI COM initialization.
Fix: Dolphin hanging forever if WASAPI wasn't supported (same fix applies to all backends)
Fix: Possibly fixed other backends that create their own audio thread by crashing or hanging dolphin when pausing the emulation.
Feature: Restart WASAPI if a device has been changed or removed.
Feature: Automatically detect when latency is too low to keep up.
Feature: Automatically detect when a the input and output (Mixer against Audio Device) got out of sync and re-initialize WASAPI as there is no other way to avoid permanent crackling once you miss a period (can easily happen when the Dolphin process hangs for a bit).
Fix: WASAPI breaking when the process was being stopped (e.g. a breakpoint in VS) and then resumed.
Fixed memory leak/crash when closing dolphin when a game was running or stopping a game, with WASAPI on.
Feature: added the ability to change DPLII, latency, device and sample rate at runtime.
Feature: WASAPI now correctly follows the volume set to the application in the Windows mixer.

I doubt cubeb will ever implement exclusive mode, as it's not very useful for them, so this is good to have.

Filoppi commented:
Input macro functions changes:
-Made using macro functions easier as they now automatically add the minimum number of required commas.
-Split them up by categories in the UI.
-Add AntiDeadzone setting, useful for games with a high deadzone and to map the mouse axis to a controller axis.
-Added BezierCurve and AntiAcceleration functions to aid the mapping of a mouse axis to a controller axis.
-Add OnPress, On Release, OnChange...
-Add meta functions like HasFocus or GetGameSpeed

techydude0713 commented on Source/Core/AudioCommon/AudioCommon.h:
```diff
@@ -16,6 +16,9 @@ class Mixer;
 
```
```suggestion
#include <mutex>

```
Add to headers in order to maintain compatibility with other platforms.

techydude0713 commented on Source/Core/DolphinQt/Settings/GeneralPane.cpp:
```diff
@@ -146,11 +146,16 @@ void GeneralPane::CreateBasic()
     if (i != 100)
-      str = QStringLiteral("%1%").arg(i);
+    {
+      if (i > 200 && std::fmod((float)i, 50.f) != 0)
+        continue;
```
You need to add ```#include <cmath>``` to maintain compatibility with other platforms.

JMC47 commented:
@dolphin-emu-bot rebuild

RinMaru commented:
no DPLII on HLE though?

Filoppi commented:
> no DPLII on HLE though?

No, this does not really touch the emulation side of the audio, just the mixing. I have no knowledge to fix that, I generally just use LLE anyway. DPLII is greatly improved though and it should work with audio stretching now (not yet, I will push the fix tonight).

Filoppi commented:
> no DPLII on HLE though?

As mentioned, I could benefit some testing

Filoppi commented:
> no DPLII on HLE though?

What quality do you set DPLII at btw?

newperson1746 commented:
Does this work transfer to linux also? I can very audibly tell audio quality degradation through dolphin vs the raw streams. For example, I can play the Mario Galaxy .asts converted to .wav (32khz) and theyll play at a perfect integer multiple (96khz) and sound excellent. Through dolphin, the high frequencies are all muddled; turns out this is because dolphin resamples to 48khz always.

Filoppi commented:
> Does this work transfer to linux also?
The audio conversion is independent from the OS, but there are a few things that are missing (or need testing) for Linux in this PR. 
By the way work on this will resume in a few weeks.

FishamanP commented:
I have great interest in DPLII and less stuttering while still maintaining low latency audio. Would love to see this PR polished and merged! Is any testing still needed from a higher-end Windows setup?

altimumdelta commented:
@FishamanP Not much will happen with this PR because it's one giant code dump that has many unrelated things mixed up. Filoppi was in a huge rush getting some code done and hasn't familiarized himself with the contribution practices, which are still rather similar than different around github or otherwise.

That said, not much may happen anytime soon, more likely never. Filoppi kept saying he has other interests and only took 1-2 weeks for this and leave forever, he did make a few hearbeats on IRC saying he didn't forget about it, how much that counts I don't know but if you really want to see sound stuff merged than the best thing you can do for now is borrow the PR and split the relevant code yourself and continue in a new PR.

Filoppi commented:
@altimumdelta thank you for you kind and understanding words... This PR is 98% finished (excluding review time) and just need to be split. I just have some other stuff going on in my life but I'm still following dolphin daily and finishing this is basically my highest priority in life as of now. So yeah I will finish it, I spent 3 months on that, I'm not letting it go to waste.
@FishamanP what needs testing is in a comment above.

altimumdelta commented:
> @altimumdelta thank you for you kind and understanding words... 

Seriously, that wasn't attitude, I was honestly stating exactly what everyone saw happened, and kept warning against. It is just that and I guess I may be using more serious tone right now, but I don't mean anything bad, I was infact on your "side" cheering and encouraging when you started. Great to see you respond very quickly.


> This PR is 98% finished (excluding review time) and just need to be split. I just have some other stuff going on in my life but I'm still following dolphin daily and finishing this is basically my highest priority in life as of now. So yeah I will finish it, I spent 3 months on that, I'm not letting it go to waste.

I think you got that a bit the other way around.
If you think it's **just** about splitting it and it'll get merged in no-time, you're most likely wrong. If you ask me, from what I can see, the longstanding team here will most likely start giving "it" (read: them, the splits) a try/test/review once you actually have made those PR splits and these, what may aswell be phantom, tests and assurances here probably won't magically carry over to the new PR, unless they have privately tested and you've privately consulted without anyone sharing that process in this discussion (happens but it would be in IRC, not totally private). So expect those split PRs to have some kind of a process to them. And it really depends on the individual, some might have totally ignored this PR, which means in their eyes the process would have only just started.
Except if bruteforce through splitting process like it's nothing and you manage to prove it's 100% tested to the rest of the team and they take your word for it, all in a week's worth of time, then great job.

Splitting PRs is probably a big PITA, I've had examples myself on a smaller scale, but it was all local WIP stuff before it ballooned. However since it's the only way to solve a situation, so it's good to just forget the negatives and roll with it, and use ways to make it easier to digest, avoid rushing (consequently helps avoid a ton of other issues), that plan of coming for ~3 weeks and leaving clearly needs to be --amended. Heh.

JosJuice commented:
@altimumdelta I would appreciate it if you stopped trying to onboard new contributors. The information you provide them with has frequently been wrong, and in this case, close to hostile.

altimumdelta commented:
> @altimumdelta I would appreciate it if you stopped trying to onboard new contributors. The information you provide them with has frequently been wrong, and in this case, close to hostile.

Fine, though I just took the opportunity to state my opinion after I've put my extra of my own time into this which I wouldn't like to go to waste, if possible.

And could you specify what information is wrong in your view? I never received a complaint before.

JosJuice commented:
> And could you specify what information is wrong in your view? I never received a complaint before.

I can't point to very many specific things since I don't keep logs beyond the current day, but there does happen to be an example from the current day: The discussion about clang-format which started 5-6 hours ago. Part of the problem is that trying to correct everything you say would take up a lot of our time, which has been a frequent pattern since when you started trying to contribute a few years ago.

altimumdelta commented:
@JosJuice I did indicate I misunderstood what was the question, and that I haven't used lint.sh much, so obviously no need to rely on my opinion there. The KDevelop having clang-format issues is the conclusion I found, tho I will investigate further, you can go check it out yourself, why would I make that up.
Why would you change the subject tho? Is there anything wrong in this topic is more fair to answer.

But Filoppi's not abandoned this, so we're fine, I'm fine, and I'll keep your advice in mind, yes I admitted at the time as well I was enthusiastic about new contributors, perhaps too enthusiastic.

Filoppi commented:
@altimumdelta I don't wanna carry on with this discussion, but personally, I feel like you don't fully know what you are talking about. Splitting this into multiple PR isn't really hard, most of the things don't overlap, I was just waiting to be done with everything before getting into the splitting part. And yes, I know that reviews take a long time, possibly months, as my older PRs are still to be merged, but there isn't much I can do about that.

Next time avoid saying stuff like:
> Filoppi was in a huge rush getting some code done and hasn't familiarized himself with the contribution practices
because that is really not what happened? I just worked on different things at the same time because I felt like it? Then I was gonna split everything at the end, that's just the way I usually do and it works, it's not really a problem for others because I will split them once I'm done, and this PR is a draft.
So yes I could have planned it better from the start, but it's only my problem.

Let's stop the discussion here please. I will finish this once I do. Thanks.

altimumdelta commented:
@Filoppi Yes you liked to do things your own way, on your own, that certainly has value such as strengthening git skills and the rest, it just wasn't expected, but we respect your decision regardless of my subjective comments. The point of the advice was to let you know it may take longer in total, and now that you're still here it's all fine.

I apologize for some of the exaggeration in the yesterday's response to FishamanP, I was merely trying to give FishamanP an IMO more realistic ETA or the possibility of it not being the expected one. It unfortunately ended up containing my late-evening frustrations that actually came from and the fact I'm quite busy/exhausted with a new PC build and OS/Software/IDE/server setups across 3 PCs around here for a couple of weeks now, after which dolphin's code should benefit more because of my focus finally moving to it the way I planned it for months! :) 

This shouldn't have happened this way, unfortunately sometimes with too many positives one gets a negative. Sorry about that.
